docs: Feature flags (#8091)
software-mansion/react-native-reanimated: packages/docs-reanimated/docs/guides/feature-flags.md

:::info

Feature flags are available since Reanimated 4.

:::

## Summary of available feature flags

| Feature flag name                                                                           |               Type                | Added in | Removed in | Default value |
| ------------------------------------------------------------------------------------------- | :-------------------------------: | :------: | :--------: | :-----------: |
| [`DISABLE_COMMIT_PAUSING_MECHANISM`](#disable_commit_pausing_mechanism)                     |  [static](#static-feature-flags)  |  4.0.0   |  &ndash;   |    `false`    |
| [`ANDROID_SYNCHRONOUSLY_UPDATE_UI_PROPS`](#android_synchronously_update_ui_props)           |  [static](#static-feature-flags)  |  4.0.0   |  &ndash;   |    `false`    |
| [`UNSTABLE_CSS_ANIMATIONS_FOR_SVG_COMPONENTS`](#unstable_css_animations_for_svg_components) |  [static](#static-feature-flags)  | nightly  |  &ndash;   |    `false`    |
| [`IOS_DYNAMIC_FRAMERATE_ENABLED`](#ios_dynamic_framerate_enabled)                           |  [static](#static-feature-flags)  | nightly  |  &ndash;   |    `false`    |

> | [`EXPERIMENTAL_MUTABLE_OPTIMIZATION`](#experimental_mutable_optimization)                   | [dynamic](#dynamic-feature-flags) | nightly  |  &ndash;   |    `false`    |

Author: These two flags actually come from `react-native-worklets` – I will move them to a separate page in `react-native-worklets` docs prior to merging, but I wanted to keep it that way for easier review.

Reviewer: I think we shouldn't publish added in nightly since we might forget to update this entry.

This feature flag works only on Android and has no effect on iOS. For more details, see [PR #7823](https://github.com/software-mansion/react-native-reanimated/pull/7823).

### `UNSTABLE_CSS_ANIMATIONS_FOR_SVG_COMPONENTS`

When enabled, CSS animations and transitions will also work for a limited set of props of several components from [`react-native-svg`](https://github.com/software-mansion/react-native-svg) library. Currently, `Circle`, `Ellipse`, `Line`, `Path` and `Rect` components are supported.

### `IOS_DYNAMIC_FRAMERATE_ENABLED`

This feature flags is supposed to improve the visual perception and perceived smoothness of computationally expensive animations. When enabled, the frame rate will be automatically adjusted for current workload of the UI thread. For instance, if the device fails to run animations in 120 fps which would usually results in irregular frame drops, the mechanism will fallback to stable 60 fps. For more details, see [PR #7624](https://github.com/software-mansion/react-native-reanimated/pull/7624).

### `EXPERIMENTAL_MUTABLE_OPTIMIZATION`

> This feature flag is supposed to speedup shared value reads on the RN runtime by reducing the number of calls to `executeOnUIRuntimeSync`. When enabled, mutables (which are the primitives behind shared values) use `Synchronizable` instead of `Shareable` to store current value of the mutable. For more details, see [PR #8080](https://github.com/software-mansion/react-native-reanimated/pull/8080).

Reviewer: ```suggestion
This feature flag is supposed to speedup shared value reads on the RN runtime by reducing the number of calls to `executeOnUIRuntimeSync`. When enabled, mutables (which are the primitives behind shared values) use `Synchronizable` state to check if they should sync with the UI Runtime. For more details, see [PR #8080](https://github.com/software-mansion/react-native-reanimated/pull/8080).
```
